### DISCUSSION TITLE
Airflow DB query takes 70% of DB cpu

### DISCUSSION THREAD

**Discussion by jiaweigoh** [apache/airflow#29698] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

Hi guys, we noticed that our Airflow DB (GCP CloudSQL Postgres) is often at 100%, with 70% consumed by a particular query. Checked via pg_stat_statements, Google's Query Insight tells us the same thing.

We are using the most basic set-up, 1 vcpu core, 3.75gb ram.

While we figure out what Airflow configs to tune, and whether to scale up our DB instance, i wanted to figure out what this query achieves. I checked that the filter conditions are indexed in Airflow DB, so its shocking for a single query type to consume so much CPU.

```
SELECT task_instance.try_number AS task_instance_try_number, task_instance.task_id AS task_instance_task_id, task_instance.dag_id AS task_instance_dag_id, task_instance.run_id AS task_instance_run_id, task_instance.start_date AS task_instance_start_date, task_instance.end_date AS task_instance_end_date, task_instance.duration AS task_instance_duration, task_instance.state AS task_instance_state, task_instance.max_tries AS task_instance_max_tries, task_instance.hostname AS task_instance_hostname, task_instance.unixname AS task_instance_unixname, task_instance.job_id AS task_instance_job_id, task_instance.pool AS task_instance_pool, task_instance.pool_slots AS task_instance_pool_slots, task_instance.queue AS task_instance_queue, task_instance.priority_weight AS task_instance_priority_weight, task_instance.operator AS task_instance_operator, task_instance.queued_dttm AS task_instance_queued_dttm, task_instance.queued_by_job_id AS task_instance_queued_by_job_id, task_instance.pid AS task_instance_pid, task_instance.executor_config AS task_instance_executor_config, task_instance.external_executor_id AS task_instance_external_executor_id, task_instance.trigger_id AS task_instance_trigger_id, task_instance.trigger_timeout AS task_instance_trigger_timeout, task_instance.next_method AS task_instance_next_method, task_instance.next_kwargs AS task_instance_next_kwargs, dag_run_1.state AS dag_run_1_state, dag_run_1.id AS dag_run_1_id, dag_run_1.dag_id AS dag_run_1_dag_id, dag_run_1.queued_at AS dag_run_1_queued_at, dag_run_1.execution_date AS dag_run_1_execution_date, dag_run_1.start_date AS dag_run_1_start_date, dag_run_1.end_date AS dag_run_1_end_date, dag_run_1.run_id AS dag_run_1_run_id, dag_run_1.creating_job_id AS dag_run_1_creating_job_id, dag_run_1.external_trigger AS dag_run_1_external_trigger, dag_run_1.run_type AS dag_run_1_run_type, dag_run_1.conf AS dag_run_1_conf, dag_run_1.data_interval_start AS dag_run_1_data_interval_start, dag_run_1.data_interval_end AS dag_run_1_data_interval_end, dag_run_1.last_scheduling_decision AS dag_run_1_last_scheduling_decision, dag_run_1.dag_hash AS dag_run_1_dag_hash 
FROM task_instance JOIN dag_run AS dag_run_1 ON dag_run_1.dag_id = task_instance.dag_id AND dag_run_1.run_id = task_instance.run_id 
WHERE task_instance.dag_id = ? AND task_instance.task_id = ? AND task_instance.run_id = ? 
 LIMIT ? FOR UPDATE
```

Thanks!

### How to reproduce

NA

### Operating System

Ubuntu

### Versions of Apache Airflow Providers

Airflow version: 2.2.3
Postgres: 9.6

### Deployment

Other Docker-based deployment

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
Converted to a discussion - because it is a discussion not an issue at this stage (please for the future use discussions if you have questions like that.

I believe that query has been optimized since 2.2.3. They were optimized for MySQL, but it could be that in some circumstances planner decides (wrongfully) to not  use the indexes. Runing explain plan on your DB might give you more insights.

Your best bet is to stop using almost 2 years old version of airflow and upgrade to latest version (2.5.1 currently).

relevant issues (there might be more):

https://github.com/apache/airflow/pull/25673: 2.3.4
https://github.com/apache/airflow/pull/25725: 2.4.0
https://github.com/apache/airflow/pull/27821: 2.5.1

Not sure which one is causing it, but my best advice is to upgrade and recheck again. This might save you (and those who look at the issues) a lot of time for investigations.

**Comment by jiaweigoh** [comment]
Thanks for the input @potiuk , sure was interesting to know about the query and indexes not being used
